hedera-node/cli-clients/src/main/java/com/hedera/services/cli/signedstate/DumpTokensSubcommand.java

```diff
@@ -0,0 +1,473 @@
+                final var sb = new StringBuilder();
+                final var b = toStructureSummaryOfJKey(sb, ojk.get());
+                if (!b) return "";
+                return c + ":" + sb + "; ";
+            });
+            return r.isEmpty() ? "" : r.substring(0, r.length() - 2);
+        }
+
+        // spotless:off
+        @NonNull
+        private static final Map<Character, Function<Token, Optional<JKey>>> KEYS = new TreeMap<>(Map.of(
+                'A', Token::adminKey,
+                'F', Token::feeScheduleKey,
+                'K', Token::kycKey,
+                'P', Token::pauseKey,
+                'S', Token::supplyKey,
+                'W', Token::wipeKey,
+                'Z', Token::freezeKey));
+        // spotless:on
+
+        @NonNull
+        private String getKeyDescription(@NonNull final BiFunction<Character, Optional<JKey>, String> map) {
+            return KEYS.entrySet().stream()
+                    .map(e -> map.apply(e.getKey(), e.getValue().apply(this)))
+                    .collect(Collectors.joining());
+        }
+    }
+
+    @NonNull
+    EnumMap<TokenType, Map<Long, Token>> gatherTokenTypes(
+            @NonNull final MerkleMapLike<EntityNum, MerkleToken> tokensStore) {
+
+        final var allTokens = new EnumMap<TokenType, Map<Long, Token>>(TokenType.class);
+        allTokens.put(TokenType.FUNGIBLE_COMMON, new HashMap<>());
+        allTokens.put(TokenType.NON_FUNGIBLE_UNIQUE, new HashMap<>());
+
+        tokensStore.forEachNode((en, mt) -> allTokens.get(mt.tokenType()).put(en.longValue(), new Token(mt)));
+        return allTokens;
+    }
+
+    void reportSummary(@NonNull Writer writer, @NonNull EnumMap<TokenType, Map<Long, Token>> allTokens) {
+        writer.writeln("=== %7d: fungible token types"
+                .formatted(allTokens.get(TokenType.FUNGIBLE_COMMON).size()));
+        writer.writeln("=== %7d: non-fungible token types"
+                .formatted(allTokens.get(TokenType.NON_FUNGIBLE_UNIQUE).size()));
+        writer.writeln("");
+    }
+
+    /** String that separates all fields in the CSV format */
+    static final String FIELD_SEPARATOR = ";";
+
+    /** String that separates sub-fields (e.g., in lists). */
+    static final String SUBFIELD_SEPARATOR = ",";
+
+    static class FieldBuilder {
+        final StringBuilder sb;
+        final String fieldSeparator;
+
+        FieldBuilder(@NonNull final String fieldSeparator) {
+            this.sb = new StringBuilder();
+            this.fieldSeparator = fieldSeparator;
+        }
+
+        void append(@NonNull final String v) {
+            sb.append(v);
+            sb.append(fieldSeparator);
+        }
+
+        @Override
+        @NonNull
+        public String toString() {
+            if (sb.length() > fieldSeparator.length()) sb.setLength(sb.length() - fieldSeparator.length());
+            return sb.toString();
+        }
+    }
+
+    static Function<Boolean, String> booleanFormatter = b -> b ? "T" : "";
```

Comment:
This convention of outputting "T" for true and "" for false is a little less explicit than I like, but probably not a big deal

Reply:
I totally get you but if you import this CSV into excel/sheets it kind of looks better - you can see the `T`s  and the ` ` stand out in a column in a way that a column of mixed `T` and `F` obscures things.  So I'd prefer to keep it this way.

Reply:
(Which is also why in the "analysis" tables of keys and stuff in this file (and also the other dumpers that do similar things) I emit a lot of blank fields instead of something else - makes the _valid_ entries stand out better.)